packages/xstate-store/src/atom.ts

```diff
@@ -110,37 +151,20 @@ export function createAtom<T>(
           link(atom, activeSub);
         }
         return atom._snapshot;
-      },
-      _update(): boolean {
-        const prevSub = activeSub;
-        const compare = options?.compare ?? Object.is;
-        activeSub = atom as InternalReadonlyAtom<T>;
-        startTracking(atom as InternalReadonlyAtom<T>);
-        try {
-          const oldValue = atom._snapshot;
-          const read = (atom: Readable<any>) => atom.get();
-          const newValue = getter(read);
-          if (oldValue === undefined || !compare(oldValue, newValue)) {
-            atom._snapshot = newValue;
-            return true;
-          }
-          return false;
-        } finally {
-          activeSub = prevSub;
-          endTracking(atom as InternalReadonlyAtom<T>);
-        }
       }
     });
   } else {
     Object.assign<BaseAtom<T>, Pick<Atom<T>, 'set'>>(atom, {
       set(valueOrFn: T | ((prev: T) => T)): void {
-        const compare = options?.compare ?? Object.is;
-        const value =
+        const prevValue = atom._snapshot;
+        const fn =
           typeof valueOrFn === 'function'
-            ? (valueOrFn as (prev: T) => T)(atom._snapshot)
-            : valueOrFn;
-        if (compare(atom._snapshot, value)) return;
-        atom._snapshot = value;
+            ? () => valueOrFn(atom._snapshot)
+            : () => valueOrFn;
+        atom._update(fn);
+        const nextValue = atom._snapshot;
+        const compare = options?.compare ?? Object.is;
+        if (compare(prevValue, nextValue)) return;
```

Comment:
a good chunk of the `set`'s implementation is already handled by `._update`. Couldn't we rely on it more? 🤔

Reply:
[357e414](https://github.com/statelyai/xstate/pull/5266/commits/357e414d250de3a3739062f00cb9b0c85fe21b40)